File: 01_数组/Demo_1.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * @author ZhaoWeiLong
 * @date 2021/3/11
 */
public class Demo_1 {

  public static void main(String[] args) {

    Integer[] num1 = new Integer[] {1, 2, 2, 1};
    Integer[] num2 = new Integer[] {2, 3, 7, 4};
    Integer[] num3 = new Integer[] {2, 3, 1, 4, 5, 8};

    System.out.println(Arrays.toString(intersection(num1, num2)));
    System.out.println(Arrays.toString(intersection2(num2, num3)));
  }

  /**
   * 求两数组交集
   *
   * @param num1
   * @param num2
   * @return
   */
  public static Integer[] intersection(Integer[] num1, Integer[] num2) {
    Map<Integer, Integer> map = new HashMap<>();
    for (int v : num1) {
      if (map.containsKey(v)) {
        map.put(v, map.get(v) + 1);
      } else {
        map.put(v, 1);
      }
    }
    ArrayList<Integer> list = new ArrayList<>();
    for (int v : num2) {
      // 存在相同，将次数-1，
      if (map.containsKey(v)) {
        map.put(v, map.get(v) - 1);
        list.add(v);
      }
    }
    return list.toArray(new Integer[0]);
  }

  /**
   * 求两数组交集 排序优化版
   *
   * @param num1
   * @param num2
   * @return
   */
  public static Integer[] intersection2(Integer[] num1, Integer[] num2) {
    List<Integer> list1 = Arrays.stream(num1).sorted().collect(Collectors.toList());
    List<Integer> list2 = Arrays.stream(num2).sorted().collect(Collectors.toList());

    ArrayList<Integer> list = new ArrayList<>();

    for (int i = 0, j = 0; i < list1.size() && j < list2.size(); ) {
      Integer v1 = list1.get(i);
      Integer v2 = list2.get(j);

      if (v1 > v2) {
        j++;
      } else if (v1 < v2) {
        i++;
      } else {
        list.add(v1);
        i++;
        j++;
      }
    }
    return list.toArray(new Integer[0]);
  }
}

File: 01_数组/Demo_1.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	//两个数组的交集

	/**
	给定两个数组，编写一个函数来计算它们的交集
	示例一：
	输入: nums1 = [1,2,2,1], nums2 = [2,2]

	输出: [2,2]

	示例二：
	输入: nums1 = [4,9,5], nums2 = [9,4,9,8,4]

	输出: [4,9]
	*/

	num1 := []int{1, 2, 2, 1}
	num2 := []int{2, 2}

	fmt.Println(intersection(num1, num2))
	fmt.Println(intersection2(num1, num2))

	//

}

func intersection(num1, num2 []int) []int {
	m := make(map[int]int)
	//将num1重复的次数添加到map中
	for _, v := range num1 {
		m[v] += 1
	}
	k := 0
	for _, v := range num2 {
		//如果相同， 将其存入num2，并将出现次数减1
		if m[v] > 0 {
			m[v] -= 1
			num2[k] = v
			k++
		}
	}
	return num2[:k]
}

//排序优化版
func intersection2(num1, num2 []int) []int {
	i, j, k := 0, 0, 0

	sort.Ints(num1)
	sort.Ints(num2)

	for i < len(num1) && j < len(num2) {
		if num1[i] > num2[j] {
			j++
		} else if num1[i] < num2[j] {
			i++
		} else {
			num1[k] = num1[i]
			i++
			j++
			k++
		}
	}

	return num1[:k]
}
